### PR TITLE
Refactor of workflows

### PR DESCRIPTION
**Pull request by Arthurvdv**
The goal is to have automatically update the (pre)release releases with new assets when a new version of the AL Language is released.

* CICD(.yml)
The `build.yml` is replaced with the `CICD.yml`, which should result in the same behavior. On a pull-request do a build+test and after a merge in the master/prerelease branch do a build+test+release.

* NewReleaseAL(.yml)
This workflow will compare the assets from the release versus the available version of the AL Language. If a new version of the AL Language is released, it wil add an extra asset to the (pre)release of the LinterCop (and replace the current/next asset). 

* DailyCheckForUpdates(.yml)
This is the trigger to execute multiple times a day the `NewReleaseAL.yml`, which in turn will check the assets with the AL Languages.

To prevent duplicate code, steps from the original `build.yml` have been refactored into various actions. This encapsulate the complexity into a single actions and could make modifications in the future simpler. The `CICD.yml` and 'NewReleaseAL.yml' share  comparisons, but without a direct dependency between the both. In this way both workflows can be adjusted for future needs.

### CODE REVIEW

**Comment by christianbraeunlich** [comment]
Hi Arthur, havent look into this in-depth. What I noticed is that, perhaps for some workflows it would make sense to execute them if the repository name corresponds to the main repository.

Unfortunately, there is no function in GitHub to control the behavior of workflows when forking repos. Most of them can only merge changes via PR anyway. A condition at the first step of the workflow whether the repository name ${{ github.repository }} == StefanMaron/lintercop would do for now.

If we want to be careful with today's resources, then it would be best to manually disable the workflows in the forked repo. The scheduled workflows would start VMs/containers on a daily basis, which is not necessary in the forked repos.

grafik

**Comment by Arthurvdv** [comment]
@christianbraeunlich, thank you for the suggestion!

If I'm not mistaken, creating a fork will default disable all the workflows, where you explicitly need to enabled them. Or are there other cases what I haven't thought about?

image

**Comment by christianbraeunlich** [comment]
> If I'm not mistaken, creating a fork will default disable all the workflows, where you explicitly need to enabled them. Or are there other cases what I haven't thought about?

You are right, makes sense! 🤔

File: .github/actions/download-artifacts/action.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        path: ${{ github.workspace }}\DownloadBuildArtifacts
+
+    - name: Rename artifacts
+      shell: pwsh
+      env:
+        SOURCE_PATH: ${{ github.workspace }}\DownloadBuildArtifacts
+        TARGET_PATH: ${{ inputs.path }}
+      run: |
+        # Get list of artifacts files
+        $artifacts = Get-ChildItem -Path $env:SOURCE_PATH -Recurse | ForEach-Object { if (!($_.PSIsContainer)) { $_}} | Where-Object { $_.Name -like 'BusinessCentral.LinterCop*.dll' -and  $_.Directory.Name -like 'BusinessCentral.LinterCop*.dll' }
```
Comment (christianbraeunlich) [inline .github/actions/download-artifacts/action.yml:24]: Also, what am I missing in the 2nd cond, why is there a .dll ending?

```suggestion
        $artifacts = Get-ChildItem -Path $env:SOURCE_PATH -Recurse -File
        | Where-Object { $_.Name -like 'BusinessCentral.LinterCop*.dll' -and  $_.Directory.Name -like 'BusinessCentral.LinterCop*.dll' }
```

**Review by christianbraeunlich — APPROVED** [review]
Splitting individual actions looks great and clear.
The ConvertTo-Version can be split into a separate file or would that be too much for now?

**Comment by StefanMaron** [comment]
Looks good, but since this is the version from main, it will probably conflict with the step I added to run tests which is currently in pre-release

**Review by StefanMaron — APPROVED** [review]

File: .github/actions/download-artifacts/action.yml (hunk 1 above)
Comment (Arthurvdv) [reply .github/actions/download-artifacts/action.yml:24]: The `actions/download-artifact@v4` download every artifact in it's own folder, where the name of the folder is the same as the artifact name.

The second condition is a extra safeguard that I only want to include these folders. In theory all the files in the `${{ github.workspace }}\DownloadBuildArtifacts` should be okay, this is just an extra precaution.

**Comment by Arthurvdv** [comment]
@christianbraeunlich, thank you for the feedback and the suggestion, very much appreciated!

@StefanMaron, I see now that I indeed forgot to also include the new `dotnet test` step you've added in the pre-release. I wanted to bring this into the repo, apart from a new release of the LinterCop itself.
I would like to discuss with you of we should run test on all AL Language versions or only focus on de `Latest` and `Pre-Release`. If a test (in the future) fails for a older version of the AL Language, I'm unsure how much work we want to put into resolving this?

**Comment by Arthurvdv** [comment]
Decided we're going for [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and for now test on all versions. If this becomes a challenge in the future, we’ll address it when the time comes.

**Pull request merged**